build-logic/src/main/kotlin/ktlint-kotlin-common.gradle.kts

```diff
@@ -8,19 +8,32 @@ plugins {
 kotlin {
     // All modules, the CLI included, must have an explicit API
     explicitApi()
+    jvmToolchain(20)
 }
 
+val targetJavaVersion = JavaVersion.VERSION_1_8
+tasks.withType<JavaCompile>().configureEach {
+    options.release.set(targetJavaVersion.majorVersion.toInt())
+}
 tasks.withType<KotlinCompile>().configureEach {
-    compilerOptions {
-        jvmTarget.set(JvmTarget.JVM_1_8)
-    }
+    kotlinOptions.jvmTarget = targetJavaVersion.toString()
 }
 
-// compileJava task and compileKotlin task jvm target compatibility should be set to the same Java version.
-// For some reason, we fallback from toolchain using, see https://github.com/pinterest/ktlint/pull/1787
-tasks.withType<JavaCompile>().configureEach {
-    sourceCompatibility = JavaVersion.VERSION_1_8.toString()
-    targetCompatibility = JavaVersion.VERSION_1_8.toString()
```

Comment:
setting `sourceCompatibility`& `targetCompatibility` can be considered legacy, `options.release` is the preferred way: 
>Although not generally advised, these options were historically used to configure the Java version during compilation.

https://docs.gradle.org/8.2/userguide/building_java_projects.html#sec:java_cross_compilation